### PR TITLE
[CA] Add CA cancellation and participants management

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/CitizenAction/CitizenActionMessageNotifier.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public function sendCancellationEmail(CitizenActionEvent $citizenActionEvent): void
+    {
+        if (!$citizenActionEvent->getCitizenAction()->isCancelled()) {
+            return;
+        }
+
+        $subscriptions = $this->registrationRepository->findByEvent($citizenActionEvent->getCitizenAction());
+
+        if (count($subscriptions) > 0) {
+            $chunks = array_chunk($subscriptions->toArray(), MailerService::PAYLOAD_MAXSIZE);
```
Comment (B-Galati) [inline src/CitizenAction/CitizenActionMessageNotifier.php]: What about calling this variable `$registrationChunks` ?

File: src/CitizenAction/CitizenActionParticipantsExporter.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        foreach ($participants as $participant) {
+            if (!array_key_exists('uuid', $participant)
+                || !array_key_exists('firstName', $participant)
+                || !array_key_exists('lastName', $participant)
+                || !array_key_exists('age', $participant)
+                || !array_key_exists('cityName', $participant)
+                || !array_key_exists('createdAt', $participant)
+                || !$participant['createdAt'] instanceof \DateTime
+            ) {
+                throw new \InvalidArgumentException('Invalid participant object given');
```
Comment (B-Galati) [inline src/CitizenAction/CitizenActionParticipantsExporter.php]: This error message should indicate which keys are missing.
I think we could use `\Symfony\Component\OptionsResolver\OptionsResolver` to help us validating these required keys and more.
If we want to stay with something simple we can always use the `array` functions to do something similar.

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 3]
```diff
@@ -82,4 +105,151 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+     * @Route("/{slug}/participants/exporter", name="app_citizen_action_export_participants")
+     * @Method("POST")
+     * @Security("is_granted('EDIT_CITIZEN_ACTION', project)")
+     */
+    public function exportParticipantsAction(Request $request, CitizenAction $citizenAction, CitizenProject $project, CitizenActionManager $citizenActionManager): Response
+    {
+        $registrations = $this->getRegistrations($request, $citizenAction, self::ACTION_EXPORT);
+
+        if (0 == $registrations->count()) {
+            return $this->redirectToRoute('app_citizen_action_list_participants', [
```
Comment (B-Galati) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: Should we add a flash message ?

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 4]
```diff
@@ -82,4 +105,151 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+    /**
+     * @Route("/{slug}/participants/imprimer", name="app_citizen_action_print_participants")
+     * @Method("POST")
+     * @Security("is_granted('EDIT_CITIZEN_ACTION', project)")
+     */
+    public function printParticipantsAction(Request $request, CitizenAction $citizenAction, CitizenProject $project, CitizenActionManager $citizenActionManager): Response
+    {
+        $registrations = $this->getRegistrations($request, $citizenAction, self::ACTION_PRINT);
+
+        if (0 == $registrations->count()) {
```
Comment (B-Galati) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: This code is really redundant. Perhaps it can be easily isolated in a private method of the controller ?

File: src/Mailer/Message/CitizenActionCancellationMessage.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        if (!$recipient instanceof EventRegistration) {
+            throw new \RuntimeException(sprintf('First recipient must be an %s instance, %s given', EventRegistration::class, get_class($recipient)));
+        }
+
+        $message = new self(
+            Uuid::uuid4(),
+            '308754',
+            $recipient->getEmailAddress(),
+            $recipient->getFirstName().' '.$recipient->getLastName(),
+            '[Projets citoyens] Une action citoyenne à laquelle vous participez vient d\'être annulée.',
```
Comment (ottaviano) [inline src/Mailer/Message/CitizenActionCancellationMessage.php]: `[Action citoyenne]`?

File: src/Mailer/Message/CitizenActionCancellationMessage.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+use AppBundle\Entity\EventRegistration;
+use Ramsey\Uuid\Uuid;
+
+final class CitizenActionCancellationMessage extends Message
+{
+    public static function create(
+        array $recipients,
+        Adherent $author,
+        CitizenAction $citizenAction,
+        \Closure $recipientVarsGenerator
```
Comment (B-Galati) [inline src/Mailer/Message/CitizenActionCancellationMessage.php]: This is fragile and un-needed (Or perhaps I am missing something :D)
Just enforce the 1st argument to be an `EventRegistration`. It's the job of the caller to do the `array_shift` stuff.

File: templates/citizen_action_manager/cancel.html.twig [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                Cette action est irréversible et enverra un e-mail d'annulation à tous les inscrits.
+            </div>
+
+            {{ form_start(form, { attr: { class: 'b__nudge--top-40'} }) }}
+
+            {{ form_errors(form) }}
+
+            <div class="form__row b__nudge--top">
+                <button type="submit" class="btn btn--blue">
+                    Oui, annuler l'actionne citoyenne
```
Comment (ottaviano) [inline templates/citizen_action_manager/cancel.html.twig]: *l'action

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * @param EventRegistration[] $recipients
+     * @param Adherent            $organizer
+     * @param string              $content
+     *
+     * @return CitizenActionContactParticipantsMessage
+     */
+    public static function create(array $recipients, Adherent $organizer, string $content): self
+    {
+        $recipients = array_values($recipients);
```
Comment (B-Galati) [inline src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: Can you add a comment to explain why is this needed ? It's to reset the key order ?

File: templates/citizen_action_manager/contact_participants.html.twig [hunk 9]
```diff
@@ -0,0 +1,62 @@
+{% extends 'citizen_action/_layout.html.twig' %}
+
+{% block page_title 'Contacter des inscrits à l\'a action citoyenne' %}
```
Comment (ottaviano) [inline templates/citizen_action_manager/contact_participants.html.twig]: *l'action

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+namespace AppBundle\Mailer\Message;
+
+use AppBundle\Entity\Adherent;
+use AppBundle\Entity\EventRegistration;
+use Ramsey\Uuid\Uuid;
+
+final class CitizenActionContactParticipantsMessage extends Message
+{
+    /**
+     * @param EventRegistration[] $recipients
```
Comment (B-Galati) [inline src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: We should keep only this line in the doc block no ?

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+     * @param Adherent            $organizer
+     * @param string              $content
+     *
+     * @return CitizenActionContactParticipantsMessage
+     */
+    public static function create(array $recipients, Adherent $organizer, string $content): self
+    {
+        $recipients = array_values($recipients);
+        $recipient = array_shift($recipients);
```
Comment (B-Galati) [inline src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: I think it would be better to let the caller decide which recipient should be the 1st one. It's not clear from the outside what it is performed in here.

File: templates/citizen_action/list_participants.html.twig [hunk 12]
```diff
@@ -2,11 +2,19 @@
 
 {% block page_title 'Particpants à l\'action citoyenne' %}
 
+{% block javascripts %}
+    <script type="text/javascript">
```
Comment (B-Galati) [inline templates/citizen_action/list_participants.html.twig]: Do we miss the call to the parent block ?

File: templates/citizen_action_manager/cancel.html.twig [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            <div class="text--body">
+                Souhaitez-vous réellement annuler l'action citoyenne {{ citizen_action.name }} ?
+            </div>
+            <div class="text--body">
+                Cette action est irréversible et enverra un e-mail d'annulation à tous les inscrits.
+            </div>
+
+            {{ form_start(form, { attr: { class: 'b__nudge--top-40'} }) }}
+
+            {{ form_errors(form) }}
```
Comment (B-Galati) [inline templates/citizen_action_manager/cancel.html.twig]: I prefer when the code is indented between `form_start` and `form_end`. What is the standard on this project ? Not sure

File: templates/citizen_action_manager/contact_participants.html.twig [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+{% block javascripts %}
+    {% import 'javascript.js.twig' as js %}
```
Comment (B-Galati) [inline templates/citizen_action_manager/contact_participants.html.twig]: Same, parent block call missing ?

File: templates/citizen_action_manager/contact_participants.html.twig [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+                            {{ msg|format('event.participants_count'|transchoice(contacts|length)) }}
+                        </div>
+                    {% endif %}
+                </div>
+
+                {{ form_start(form) }}
+
+                    {{ form_errors(form.message) }}
+
+                    <input type="hidden" name="contacts" value="{{ contacts|json_encode }}" />
```
Comment (B-Galati) [inline templates/citizen_action_manager/contact_participants.html.twig]: This could be done in the FormType no ? With the `buildView...` methods no ? I guess it's not a big deal to keep it here but it would be at least better to dynamically set the name of the input no ?

File: front/pages/manage_participants.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    function toggleButtons(value) {
+        if (null != contactButton) contactButton.disabled = !value;
+        if (null != exportButton) exportButton.disabled = !value;
+        if (null != printButton) printButton.disabled = !value;
+    }
+
+    // Toggle each 'selection' checkbox
+    function toggleCheckboxs(value) {
+        const selectCheckboxes = (element) => {
+            element.checked = value;
```
[the remark this answers is not in the document]
Comment (B-Galati) [reply front/pages/manage_participants.js]: What a trick :D

File: front/pages/manage_participants.js [hunk 17]
```diff
@@ -0,0 +1,132 @@
+/*
+ * Manage the participants of event/citizen action
+ */
+export default() => {
+    const allCheckbox = dom('#members-check-all');
+    const selectionCheckboxes = document.getElementsByName('members[]');
```
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply front/pages/manage_participants.js]: *CheckboxEs :)

File: translations/messages.fr.yml [hunk 18]
```diff
@@ -96,6 +96,7 @@ citizen_action.registration.success: Votre inscription est confirmée.
 citizen_action.cancel.success: L'action citoyenne a bien été annulée.
 citizen_action.contact.none: Vous n'avez sélectionné aucun inscrit à contacter.
 citizen_action.contact.success: Félicitations, votre message a bien été envoyé aux inscrits sélectionnés.
+citizen_action.export.none: Aucun inscrit à l'initiative citoyenne n'a pas été trouvé.
```
Comment (ottaviano) [inline translations/messages.fr.yml]: `n'a été trouvé` sans `pas` à cause de `aucun`

File: src/Mailer/Message/CitizenActionCancellationMessage.php (hunk 6 above)
Comment (Remg) [reply src/Mailer/Message/CitizenActionCancellationMessage.php]: About the closure generating the recipient vars: actually this is not needed and will be refactored in the `email-templates` PR, i suggest you leave this as is in this PR.

File: src/CitizenAction/CitizenActionContactParticipantsCommand.php [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+use AppBundle\Entity\Adherent;
+use Symfony\Component\Validator\Constraints as Assert;
+
+class CitizenActionContactParticipantsCommand
+{
+    /** @var Adherent[] */
+    private $recipients;
+
+    /** @var Adherent */
```
Comment (Remg) [inline src/CitizenAction/CitizenActionContactParticipantsCommand.php]: this docblock is not needed

File: src/CitizenAction/CitizenActionContactParticipantsCommand.php [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return $this->recipients;
+    }
+
+    public function getSender(): Adherent
+    {
+        return $this->sender;
+    }
+
+    public function setMessage(?string $message)
```
Comment (Remg) [inline src/CitizenAction/CitizenActionContactParticipantsCommand.php]: missing return type

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php (hunk 10 above)
Comment (Remg) [reply src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: @B-Galati: if we specify a docblock, then it must be complete.

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+use Ramsey\Uuid\Uuid;
+
+final class CitizenActionContactParticipantsMessage extends Message
+{
+    /**
+     * @param EventRegistration[] $recipients
+     * @param Adherent            $organizer
+     * @param string              $content
+     *
+     * @return CitizenActionContactParticipantsMessage
```
Comment (Remg) [inline src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: `@return self` could be better, just my opinion \o/

File: templates/citizen_action_manager/cancel.html.twig [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+                Annuler l'action citoyenne
+            </h2>
+            <div class="text--body">
+                Souhaitez-vous réellement annuler l'action citoyenne {{ citizen_action.name }} ?
+            </div>
+            <div class="text--body">
+                Cette action est irréversible et enverra un e-mail d'annulation à tous les inscrits.
+            </div>
+
+            {{ form_start(form, { attr: { class: 'b__nudge--top-40'} }) }}
```
Comment (Remg) [inline templates/citizen_action_manager/cancel.html.twig]: `{ class: 'b__nudge--top-40' }`
🏃

File: templates/citizen_action_manager/cancel.html.twig [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+                {{ form_errors(form) }}
+
+                <div class="form__row b__nudge--top">
+                    <button type="submit" class="btn btn--blue">
+                        Oui, annuler l'action citoyenne
+                    </button>
+                    <a href="{{ path('app_citizen_action_show', { slug: citizen_action.slug}) }}"
```
Comment (Remg) [inline templates/citizen_action_manager/cancel.html.twig]: `{ slug: citizen_action.slug }`
🏃

File: src/Mailer/Message/CitizenActionContactParticipantsMessage.php (hunk 11 above)
Comment (Remg) [reply src/Mailer/Message/CitizenActionContactParticipantsMessage.php]: I agree with you, this system is not very clear.
The abstract Message class handles this logic (waiting for a **main** recipient + an array of recipients), this could be refactored but out of scope imho

File: src/Controller/EnMarche/CitizenActionManagerController.php (hunk 4 above)
Comment (illusionOfParadise) [reply src/Controller/EnMarche/CitizenActionManagerController.php]: It's not useful, if we create a new method it will need a new `if` to do or not the redirection.

File: front/pages/manage_participants.js [hunk 24]
```diff
@@ -0,0 +1,132 @@
+/*
+ * Manage the participants of event/citizen action
+ */
+export default() => {
+    const allCheckboxes = dom('#members-check-all');
+    const selectionCheckboxes = document.getElementsByName('members[]');
```
Comment (ottaviano) [inline front/pages/manage_participants.js]:
```js 
findAll(document, 'input[name="members[]"]')
``` 
(cf `dom.js` file)

File: front/pages/manage_participants.js [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+        if (null != printButton) printButton.disabled = !value;
+    }
+
+    // Toggle each 'selection' checkbox
+    function toggleCheckboxs(value) {
+        const selectCheckboxes = (element) => {
+            element.checked = !!value;
+        };
+
+        selectionCheckboxes.forEach(selectCheckboxes);
```
Comment (ottaviano) [inline front/pages/manage_participants.js]: can you try to avoid the constant declaration? 
I'm wondering for a visibility of `value` variable in this case :)
```js
selectionCheckboxes.forEach((element) => {
  element.checked = !!value;
});
```

File: front/pages/manage_participants.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    // Bind the 'select all' checkbox
+    const bindAllCheckboxes = (event) => {
+        if (allCheckboxes.checked) {
```
Comment (ottaviano) [inline front/pages/manage_participants.js]:
```js
const bindAllCheckboxes = (event) => {
  toggleCheckboxes(allCheckboxes.checked);
  toggleButtons(allCheckboxes.checked);
};
```

File: front/pages/manage_participants.js [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    };
+
+    const bindAllSelectionCheckboxes = (element) => {
+        on(element, 'change', toggle);
+        allCheckboxes.dispatchEvent(new Event('change'));
+    };
+
+    // Bind each 'selection' checkbox
+    selectionCheckboxes.forEach(bindAllSelectionCheckboxes);
```
Comment (ottaviano) [inline front/pages/manage_participants.js]:
```js
selectionCheckboxes.forEach((element) => {
  on(element, 'change', toggle);
  allCheckboxes.dispatchEvent(new Event('change'));
});
```

File: front/pages/manage_participants.js [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+
+    const exportMembers = (event) => {
+        const members = [];
+        const addMember = (element) => {
```
Comment (ottaviano) [inline front/pages/manage_participants.js]: can you refacto this stuff for selecting the checked members to avoid the duplication with:
```
const contactMembers = (event) => {
 +        const members = [];
 +        const addMember = (element) => {
 +            if (element.checked) {
 +                members.push(element.value);
 +            }
 +        };
 +
 +        selectionCheckboxes.forEach(addMember);
 +
 +        contactSelection.value = JSON.stringify(members);
 +    };
```

and this 
```
 const printMembers = (event) => {
 +        const members = [];
 +        const addMember = (element) => {
 +            if (element.checked) {
 +                members.push(element.value);
 +            }
 +        };
 +
 +        selectionCheckboxes.forEach(addMember);
 +
 +        printSelection.value = JSON.stringify(members);
 +    };
```

:)

File: src/CitizenAction/CitizenActionCommandHandler.php [hunk 29]
```diff
@@ -46,4 +46,15 @@ public function handleUpdate(CitizenActionCommand $command, CitizenAction $actio
 
         return $action;
     }
+
+    public function handleCancel(CitizenAction $action, CitizenActionCommand $command)
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionCommandHandler.php]: what about passing `Adherent $author` instead of `CitizenActionCommand`?

File: src/CitizenAction/CitizenActionContactParticipantsCommand.php [hunk 30]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace AppBundle\CitizenAction;
+
+use AppBundle\Entity\Adherent;
+use Symfony\Component\Validator\Constraints as Assert;
+
+class CitizenActionContactParticipantsCommand
+{
+    /** @var Adherent[] */
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionContactParticipantsCommand.php]: use multiline comment pls

File: src/CitizenAction/CitizenActionContactParticipantsCommand.php [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private $message;
+
+    /**
+     * @Assert\NotBlank
+     * @Assert\Length(max=80)
+     */
+    private $subject;
+
+    public function __construct(array $recipients, Adherent $sender, string $subject = null, string $message = null)
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionContactParticipantsCommand.php]:
```php
function __construct(Adherent $sender, array $recipients, ...)
```

File: src/CitizenAction/CitizenActionMessageNotifier.php [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            Events::CITIZEN_ACTION_CANCELLED => ['sendCancellationEmail'],
+        ];
+    }
+
+    private function createCancelMessage(array $registered, CitizenAction $citizenAction, Adherent $host): CitizenActionCancellationMessage
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionMessageNotifier.php]: `$host` (Author) is nullable and optional in `CitizenActionEvent` constructor

File: src/CitizenAction/CitizenActionMessageNotifier.php [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+        $this->registrationRepository = $registrationRepository;
+    }
+
+    public function sendCancellationEmail(CitizenActionEvent $citizenActionEvent): void
+    {
+        if (!$citizenActionEvent->getCitizenAction()->isCancelled()) {
+            return;
+        }
+
+        $subscriptions = $this->registrationRepository->findByEvent($citizenActionEvent->getCitizenAction());
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionMessageNotifier.php]: you should use a variable for citizenAction

File: src/CitizenAction/CitizenActionParticipantsExporter.php [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace AppBundle\CitizenAction;
+
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class CitizenActionParticipantsExporter
+{
+    public function export(array $participants): string
+    {
+        $handle = fopen('php://memory', 'r+');
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionParticipantsExporter.php]: what about do this with a template twig
```twig
Col1;Col2;Col3;Col4
{% for row in data %}
    {{ row.field1 }};{{ row.field2 }};{{ row.field3 }};{{ row.field4 }}
{% endfor %}
``` 

```php
public function export(array $participants): string
{
  return $this->templating->renderView(
    'view_path.twig',
    ['data' => $participants]
  );
}
```


maybe directly in `*Controller::exportParticipantsAction`?

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 35]
```diff
@@ -82,4 +105,156 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+    /**
+     * @Route("/{slug}/annuler", name="app_citizen_action_manager_cancel")
+     * @Method("GET|POST")
+     * @Security("is_granted('CANCEL_CITIZEN_ACTION', project)")
+     */
+    public function cancelAction(Request $request, CitizenAction $action, CitizenProject $project): Response
+    {
+        $command = CitizenActionCommand::createFromCitizenAction($action);
+
+        $form = $this->createForm(FormType::class);
```
Comment (ottaviano) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: you can chain the calls
```php
$this
  ->createForm(FormType::class)
  ->handleRequest($request)
;
```

File: src/CitizenAction/CitizenActionCommandHandler.php [hunk 36]
```diff
@@ -46,4 +46,15 @@ public function handleUpdate(CitizenActionCommand $command, CitizenAction $actio
...
+    {
+        $action->cancel();
+
+        $this->manager->flush();
+
+        $this->dispatcher->dispatch(Events::CITIZEN_ACTION_CANCELLED, new CitizenActionEvent($action, $command->getAuthor()));
+
+        return $action;
```
Comment (ottaviano) [inline src/CitizenAction/CitizenActionCommandHandler.php]: not need to return this value

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 37]
```diff
@@ -82,4 +105,156 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+
+            return $this->redirectToRoute('app_citizen_action_list_participants', [
+                'slug' => $citizenAction->getSlug(),
+            ]);
+        }
+
+        $participants = $citizenActionManager->populateRegistrationWithAdherentsInformations($registrations);
+        $exported = $this->get(CitizenActionParticipantsExporter::class)->export($participants);
+
+        return new Response($exported, 200, [
```
Comment (ottaviano) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: SnnapyResponse ?

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 38]
```diff
@@ -82,4 +105,156 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+        );
+    }
+
+    private function getRegistrations(Request $request, CitizenAction $citizenAction, string $action): EventRegistrationCollection
+    {
+        if (!in_array($action, self::ACTIONS)) {
+            throw new \InvalidArgumentException("Action '$action' is not allowed.");
+        }
+
+        if (!$this->isCsrfTokenValid(sprintf('citizen_action.%s_participants', $action), $request->request->get('token'))) {
```
Comment (ottaviano) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: how you creat the `citizen_action.export_participants` token id?

File: src/Controller/EnMarche/CitizenActionManagerController.php [hunk 39]
```diff
@@ -82,4 +105,156 @@ public function editAction(Request $request, CitizenProject $project, CitizenAct
...
+
+        if (!$uuids = json_decode($request->request->get(sprintf('%ss', $action)), true)) {
+            if (self::ACTION_CONTACT === $action) {
+                $this->addFlash('info', $this->get('translator')->trans('citizen_action.contact.none'));
+            }
+
+            return new EventRegistrationCollection();
+        }
+
+        $repository = $this->getDoctrine()->getRepository(EventRegistration::class);
```
Comment (ottaviano) [inline src/Controller/EnMarche/CitizenActionManagerController.php]: the variable is useless

File: src/Repository/EventRegistrationRepository.php [hunk 40]
```diff
@@ -41,6 +41,13 @@ public function findByUuidAndEvent(BaseEvent $event, array $uuids): array
         return $this->findBy(['event' => $event, 'uuid' => $uuids]);
     }
 
+    public function getByUuidAndEvent(BaseEvent $event, array $uuids): EventRegistrationCollection
```
Comment (ottaviano) [inline src/Repository/EventRegistrationRepository.php]: if the first argument is `Event` your method should be named as `getByEventAndUuids`

File: templates/citizen_action/list_participants.html.twig [hunk 41]
```diff
@@ -17,30 +26,29 @@
...
+                        <form method="post" action="{{ path("app_citizen_action_print_participants", { slug: citizen_action.slug, project_slug: citizen_action.citizenProject.slug }) }}">
+                            <input type="hidden" name="token" value="{{ csrf_token('citizen_action.print_participants') }}" id="members-print-token" />
+                            <input type="hidden" name="prints" value="" id="members-print-selection" />
+                            <button id="members-print-button" class="btn" disabled="disabled">Imprimer</button>
+                        </form>
+                    </div>
+                </div>
+            {% endif %}
+            
+            {% include 'citizen_action/_participants_list.html.twig' with { participants: participants, is_host: is_host } only %}
```
Comment (ottaviano) [inline templates/citizen_action/list_participants.html.twig]: Included templates have access to the variables of the active context, not need `with`

File: templates/citizen_action_manager/cancel.html.twig [hunk 42]
```diff
@@ -0,0 +1,36 @@
...
+            <h2 class="b__nudge--bottom-large text--medium">
+                Annuler l'action citoyenne
+            </h2>
+            <div class="text--body">
+                Souhaitez-vous réellement annuler l'action citoyenne {{ citizen_action.name }} ?
+            </div>
+            <div class="text--body">
+                Cette action est irréversible et enverra un e-mail d'annulation à tous les inscrits.
+            </div>
+
```
Comment (ottaviano) [inline templates/citizen_action_manager/cancel.html.twig]: it's me or made a new page for validating one action it's useless and old school? 
what about a confirm modal?

File: templates/citizen_action_manager/print_participants.html.twig [hunk 43]
```diff
@@ -0,0 +1,6 @@
+{% extends 'base_print.html.twig' %}
+
+{% block content %}
+    <p class="text--bold">Liste des participants</p>
+    {% include 'citizen_action/_participants_list.html.twig' with { participants: participants, is_host: true } only %}
```
Comment (ottaviano) [inline templates/citizen_action_manager/print_participants.html.twig]: useless `participants: participants`

File: src/CitizenAction/CitizenActionCommandHandler.php (hunk 29 above)
Comment (Remg) [inline src/CitizenAction/CitizenActionCommandHandler.php]: missing return type

File: src/Controller/EnMarche/EventManagerController.php [hunk 44]
```diff
@@ -212,10 +212,8 @@ private function getRegistrations(Request $request, Event $event, string $action
             return [];
         }
 
-        $repository = $this->getDoctrine()->getRepository(EventRegistration::class);
-
         try {
-            $registrations = $repository->findByUuidAndEvent($event, $uuids);
+            $registrations = $this->getDoctrine()->getRepository(EventRegistration::class)->findByUuidAndEvent($event, $uuids);
```
Comment (ottaviano) [inline src/Controller/EnMarche/EventManagerController.php]: `getByEventAndUuid `?

File: src/CitizenAction/CitizenActionMessageNotifier.php (hunk 32 above)
Comment (illusionOfParadise) [reply src/CitizenAction/CitizenActionMessageNotifier.php]: Done with `CitizenActionCancelledEvent`.

File: templates/citizen_action_manager/print_participants.html.twig (hunk 43 above)
Comment (illusionOfParadise) [reply templates/citizen_action_manager/print_participants.html.twig]: Here it's useful, theses variables are not introduced in parent template.

File: src/Controller/EnMarche/CitizenActionManagerController.php (hunk 38 above)
Comment (illusionOfParadise) [reply src/Controller/EnMarche/CitizenActionManagerController.php]: In the template.

File: templates/citizen_action_manager/cancel.html.twig (hunk 42 above)
Comment (illusionOfParadise) [reply templates/citizen_action_manager/cancel.html.twig]: This is done to follow the events functionality.

File: src/CitizenAction/CitizenActionParticipantsExporter.php (hunk 34 above)
Comment (illusionOfParadise) [reply src/CitizenAction/CitizenActionParticipantsExporter.php]: It should be better to create a PR dedicated to this refacto because there is another class `Exporter` with the same logic.
